_posts/2018-07-10-Play Framework-in-Scala–Action-Composition.md

```diff
@@ -0,0 +1,256 @@
+
+The left type parameter of an ActionBuilder constrains the signature of the block function in the invokeBlock method:
+
+```scala
+trait ActionBuilder[+R[_], B] extends ActionFunction[Request, R] {
+  def invokeBlock[A](request: Request[A],
+                     block: R[A] => Future[Result]): Future[Result]
+```
+
+On the custom LoggingActionBuilder we used type Request, where we log a request and then apply the block function to that same request object:
+
+```scala
+def invokeBlock[A](request: Request[A],
+                   block: Request[A] => Future[Result]): Future[Result] = {
+  Logger.debug(s"Just received the following request: $request")
+  block(request)
+}
+```
+
+Often it is convenient that a custom ActionBuilder, via its invokeBlock method, 
+takes a request, does something with it, builds an instance of another type, and <strong>only then</strong> applies the block function to that new instance.
+Authentication of requests is a good example. Here, a particular request, if authenticated, is normally associated with a User. It is therefore useful to obtain the corresponding instance of User and pass it along, meaning, applying the block function to that instance.
+This is useful because now the Actions that are build through this ActionBuilder may count on the existence of an instance of User instead of a raw request. 
+This of course changes the signature of the block function that must be provided for the posterior creation of each Action.
+A naive AuthenticationActionBuilder could be:
+
+```scala
+class AuthenticatedAction[B](val parser: BodyParser[B])
+                            (implicit val executionContext: ExecutionContext) 
+                            extends ActionBuilder[AuthenticatedUser, B] {
+
+override def invokeBlock[A](request: Request[A],
+                            block: AuthenticatedUser[A] => Future[Result]): Future[Result] = {
+ val userId: Option[Int] = request.headers.get("user_id")
+ val password: Option[String] = request.headers.get("password")
+
+ val user: Option[User] = DataBase.validate(userId, password)
+
+ user.fold {
+   Future.successful(BadRequest("Authentication failed"))
+ }{ user =>
+   val authenticatedUser = AuthenticatedUser(user, request)
+   block(authenticatedUser)
+ }
+ }
+}
+```
+
+Notice above, that the type parameter of the ActionBuilder changed to  <code>AuthenticatedUser</code> and correspondingly the signature of the block function on the invokeBlock method is <small><code> block: AuthenticatedUser[A] => Future[Result] </code></small>. 
+It is up to us to define type AuthenticatedUser[_]. In this case, it could something along:
+```scala
+case class AuthenticatedUser[A](user: User, request: Request[A]) extends WrappedRequest(request)
+```
+
+Where type <code>User</code> is our internal representation of a user.
+AuthenticatedUser[_] must allow a type constructor, as demanded by the signature of the invokeBlock method of the trait ActionBuilder. 
```

Comment:
The formatting is weird.

![image](https://user-images.githubusercontent.com/34140210/43081077-17bdde4a-8e89-11e8-9524-a3a95aac2303.png)